bundles/org.openhab.binding.dwdpollenflug/src/main/java/org/openhab/binding/dwdpollenflug/internal/dto/DWDPollenflug.java

```diff
@@ -0,0 +1,115 @@
+
+import static org.openhab.binding.dwdpollenflug.internal.DWDPollenflugBindingConstants.*;
+
+import java.text.SimpleDateFormat;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+
+/**
+ * The {@link DWDPollenflug} class is internal DWD data structure.
+ *
+ * @author Johannes DerOetzi Ott - Initial contribution
+ */
+@NonNullByDefault
+public class DWDPollenflug {
+    private static final SimpleDateFormat FORMATTER = new SimpleDateFormat(DATE_PATTERN);
+
+    private final Date created = new Date();
+
+    private final @Nullable Date nextUpdate;
+
+    private final @Nullable Date lastUpdate;
+
+    private final Map<String, String> properties;
+
+    private final Map<String, State> channels = new HashMap<>();
+
+    private final Map<Integer, DWDRegion> regions = new HashMap<>();
+
+    public DWDPollenflug(DWDPollenflugJSON json) {
+        nextUpdate = json.getNextUpdate();
+        lastUpdate = json.getLastUpdate();
+
+        properties = initProperties(json);
+
+        for (DWDRegionJSON regionJSON : json.getRegions()) {
+            DWDRegion region = new DWDRegion(regionJSON);
+            regions.put(region.getRegionID(), region);
+        }
+
+        parseChannels(json);
+    }
+
+    private synchronized Map<String, String> initProperties(DWDPollenflugJSON json) {
+        Map<String, String> map = new HashMap<>();
+
+        map.put(PROPERTY_NAME, json.getName());
+        map.put(PROPERTY_SENDER, json.getSender());
+        map.put(PROPERTY_REFRESHED, FORMATTER.format(created));
+
+        if (nextUpdate != null) {
+            map.put(PROPERTY_NEXT_UPDATE, FORMATTER.format(nextUpdate));
+        }
+
+        if (lastUpdate != null) {
+            map.put(PROPERTY_LAST_UPDATE, FORMATTER.format(lastUpdate));
+        }
+
+        return Collections.unmodifiableMap(map);
+    }
+
+    public Map<String, String> getProperties() {
+        return properties;
+    }
+
+    private void parseChannels(DWDPollenflugJSON json) {
+        createChannel(CHANNEL_REFRESHED, created);
+        createChannel(CHANNEL_NEXT_UPDATE, nextUpdate);
+        createChannel(CHANNEL_LAST_UPDATE, lastUpdate);
+    }
+
+    private void createChannel(String subchannel, @Nullable Date date) {
+        final String channelName = CHANNEL_UPDATES + "#" + subchannel;
+        if (date == null) {
+            channels.put(channelName, UnDefType.NULL);
+        } else {
+            ZonedDateTime zoned = ZonedDateTime.ofInstant(date.toInstant(), ZoneId.systemDefault());
+            channels.put(channelName, new DateTimeType(zoned));
+        }
+    }

```

Comment:
You're right the parseChannels is a bit overkill ;) but I think the createChannel can be left there because the channel name, the null check and the date conversion is three times is in my opinion complex enough redudant code.

Reply:
Please check my fix!

Reply:
Changed NP-handling in DTOs, so I changed it now.

Reply:
Refactored whole DTO closing this conversation, because of huge changes.